scipy/special/xsf/zeta.h

```diff
@@ -1,10 +1,279 @@
+ */
+
 #pragma once
 
+#include "config.h"
+#include "error.h"
+
+#include "cephes/const.h"
 #include "cephes/zeta.h"
 #include "cephes/zetac.h"
 
 namespace xsf {
 
+namespace detail {
+
+    /* Log of absolute value of expansion coefficients for Euler-Maclaurin
+     * summation formula.
+     * log(|B2k / (2k)!|) */
+    constexpr double zeta_em_log_abs_coeff_lookup[] = {
+	-2.4849066497880004,
+	-6.579251212010101,
+	-10.31692083029347,
+	-14.005800284407405,
+	-17.68462940266784,
+	-21.361131560073222,
+	-25.037070502911423,
+	-28.712870599846948,
+	-32.388636197522295,
+	-36.06439319366539,
+	-39.740148041995184,
+	-43.41590235365616,
+	-47.091656531181485,
+	-50.76741067517639,
+	-54.44316481078909,
+	-58.11891894430628,
+	-61.79467307729959,
+	-65.47042721016194,
+	-69.14618134299154,
+	-72.82193547581296,
+	-76.49768960863234,
+	-80.1734437414512,
+	-83.84919787426993,
+	-87.52495200708863,
+	-91.20070613990733,
+	-94.87646027272602,
+	-98.55221440554472,
+	-102.2279685383634,
+	-105.9037226711821,
+	-109.57947680400078,
+	-113.25523093681947,
+	-116.93098506963817,
+	-120.60673920245685,
+	-124.28249333527555,
+	-127.95824746809424,
+	-131.63400160091294,
+	-135.30975573373163,
+	-138.9855098665503,
+	-142.661263999369,
+	-146.3370181321877,
+	-150.0127722650064,
+	-153.6885263978251,
+	-157.36428053064375,
+	-161.04003466346245,
+	-164.71578879628115,
+	-168.39154292909984,
+	-172.06729706191854,
+	-175.74305119473723,
+	-179.4188053275559,
+	-183.0945594603746
+    };
+
+    // Complex log of expansion coefficients for Euler-Maclaurin summation formula.
+    XSF_HOST_DEVICE inline std::complex<double> zeta_em_log_coeff(std::size_t n) {
+	std::complex<double> J(0.0, 1.0);
+	std::complex<double> result;
+	if (n < 50) {
+	    result = zeta_em_log_abs_coeff_lookup[n];
+	} else {
+	    // asymptotic formula
+	    result = std::log(2.0) - 2.0*(n + 1)*std::log(2*M_PI);
+	}
+	if (n % 2 == 1) {
+	    /* B_{2n}/(2n)! is negative for even n. This contributes a term
+	     * pi*i when taking the log. */
+	    result += M_PI * J;
+	}
+	return result;
+    }
+
+    /* Compute riemann_zeta for complex input z using the Euler-Maclaurin formula.
+     * Computation of individual terms in expansion are logarithmized to avoid
+     * overflow. TODO: only logarithmize when necessary. */
+    XSF_HOST_DEVICE inline std::complex<double> zeta_euler_maclaurin(std::complex<double> z) {
+	if (z.real() == 1.0) {
+	    return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
+	}
+	std::size_t n = static_cast<std::size_t>(std::max(std::abs(z.imag()) / 4.0, 50.0));
+	std::size_t m = n;
+	std::complex<double> result = 0.0;
+	for (std::size_t i = 1; i < n; i++) {
+	    std::complex<double> term = std::pow(static_cast<double>(i), -z);
+	    result += term;
+	    if (std::abs(term) / std::abs(result) <= std::numeric_limits<double>::epsilon()) {
+		return result;
+	    }
```

Comment:
Good question. I borrowed this from the `cephes` real implementation. Strictly speaking, the partial sum doesn't converge when `z.real <= 1`, so we would hope this early termination only kicks in when the sum is convergent. When `0 < z.real <= 1` the terms still go to zero though, and we could terminate even though the sum doesn't converge.

To be safe, I'll change it to

```C++
if (z.real() > 1 && std::abs(term) / std::abs(result) <= std::numeric_limits<double>::epsilon()) {
	     	return result;
	    }
```

Reply:
I think it could potentially be a problem for `z` close to  1, but we don’t use Euler-Maclaurin there. Although, I’ve just discovered that borwein’s algorithm isn’t doing great near `z = 1` either.